src/agent/program.cpp

```diff
@@ -0,0 +1,124 @@
+ *    All rights reserved.
+ *
+ *    Redistribution and use in source and binary forms, with or without
+ *    modification, are permitted provided that the following conditions are met:
+ *    1. Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *    2. Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *    3. Neither the name of the copyright holder nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ *    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *    AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *    IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *    ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *    LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *    CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *    SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *    INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *    CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *    ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *    POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   The file implements the OTBR Agent.
+ */
+
+#define OTBR_LOG_TAG "AGENT"
+
+#include "agent/program.hpp"
+#include <common/mainloop_manager.hpp>
+#include "common/code_utils.hpp"
+
+namespace otbr {
+
+bool Program::sShouldTerminate = false;
+
+Program::Program(Ncp::ControllerOpenThread &aOpenThread)
+    : mBorderAgent(aOpenThread)
+#if OTBR_ENABLE_OPENWRT
+    , mUbusAgent(aOpenThread)
+#endif
+#if OTBR_ENABLE_REST_SERVER
+    , mRestWebServer(aOpenThread)
+#endif
+#if OTBR_ENABLE_DBUS_SERVER
+    , mDBusAgent(aOpenThread)
+#endif
+#if OTBR_ENABLE_VENDOR_SERVER
+    , mVendorServer(aOpenThread)
+#endif
+{
+}
+
+void Program::Init(void)
+{
+    mBorderAgent.Init();
+#if OTBR_ENABLE_OPENWRT
+    mUbusAgent.Init();
+#endif
+#if OTBR_ENABLE_REST_SERVER
+    mRestWebServer.Init();
+#endif
+#if OTBR_ENABLE_DBUS_SERVER
+    mDBusAgent.Init();
+#endif
+#if OTBR_ENABLE_VENDOR_SERVER
+    mVendorServer.Init();
+#endif
+}
+
+otbrError Program::Run(void)
+{
+    otbrError error = OTBR_ERROR_NONE;
+
+    otbrLogInfo("Border router agent started.");
+    // allow quitting elegantly
+    signal(SIGTERM, HandleSignal);
+
+    while (!sShouldTerminate)
+    {
+        otbr::MainloopContext mainloop;
+        int                   rval;
+
+        mainloop.mMaxFd   = -1;
+        mainloop.mTimeout = kPollTimeout;
```

Comment:
Should we consider remove this default timeout? If there's no no scheduled tasks, select() blocks until an event occurs.

Reply:
I agree. But may leave for another PR.